WordPress/gutenberg
# Global styles: add custom CSS panel to site editor

## What?
Adds a custom CSS input option to global styles in the site editor.

## Why?
Because this option was available to users in customizer, but there is no option to add arbitrary CSS via the site editor.
Part of: #30142

## How?

- Adds a `css` property to [theme.json styles](https://github.com/WordPress/gutenberg/pull/46255) -  `styles.css`
- Saves the global custom CSS to the `wp_global_styles` CPT
- In the site editor the content of `styles.css` is added to a separate style sheet. Initially, it was just appended to the end of the existing global styles style sheet, but the issue with this approach was that any invalid CSS entered caused all global styles to stop working as the styles are parsed with js. Having a separate style sheet means any invalid CSS entered only breaks the custom CSS and all other global styles work as expected. It seems like just appending it to the end of the global styles style sheet works for the post editor and frontend though.

## Testing Instructions

- In the site editor open the global styles panel
- Go to Custom CSS option at the bottom of the panel
- Enter your favorite CSS and save
- Check the CSS is applied in site editor, post editor and front end
- Add a `styles.css` property to your theme.json file and make sure that it displays as expected. Then add some user custom CSS, and make sure the theme.json setting displays in the input box  and make sure that this can be removed or overriden in site and page editor and front end
- Add a `styles.css` property to the core theme.json, and make sure this is correctly overridden by the themes theme.json setting and also the user custom CSS

## Known issue

Currently, any changes to the custom CSS that are not valid CSS will cause all the custom CSS to stop applying until the CSS is made valid again. Because of the way that CSS is parsed and rewritten for the site editor this is going to be difficult to fix so will be handled in a follow up.

## Screenshots or screencast 

https://user-images.githubusercontent.com/3629020/205171743-47a3f1ad-d386-4e22-b70f-5a2d732b470d.mp4
